ios/RNCardConnectReactLibrary.m

```diff
@@ -37,15 +103,216 @@ - (dispatch_queue_t)methodQueue
     }];
 }
 
+RCT_EXPORT_METHOD(discoverDevice) {
+
+    if (_swiper.connectionState == BMSSwiperConnectionStateSearching) {
+        [_swiper cancelFindDevices];
+    }
+    else if (_swiper.connectionState == BMSSwiperConnectionStateConnecting) {
+        [_swiper releaseDevice];
+    }
+
+    [self.swiper findDevices];
+}
+
+RCT_EXPORT_METHOD(activateDevice) {
+
+    if (self.restartReaderBlock) {
+        self.restartReaderBlock();
+        self.restartReaderBlock = nil;
+    }
+}
+
+RCT_EXPORT_METHOD(cancelTransaction) {
+
+    [self.swiper cancelTransaction];
+}
+
+RCT_EXPORT_METHOD(connectToDevice:(NSString *)uuid) {
+
+    if (_swiper.connectionState == BMSSwiperConnectionStateSearching) {
+        [_swiper cancelFindDevices];
+    }
+
+    NSUUID *converted = [[NSUUID alloc] initWithUUIDString:uuid];
+
+    [_swiper connectToDevice:converted mode:BMSCardReadModeSwipeDipTap];
+
+    self.isConnecting = true;
+}
+
+- (void) debug:(NSString *)message {
+
+    if (self.enableDebugging) {
+        [self sendEventWithName:@"BoltOnLogUpdate" body:@{@"debugMessage": message}];
+    }
+}
+
+- (void)swiper:(BMSSwiperController *)swiper configurationProgress:(float)progress {
+
+    [self sendEventWithName:@"BoltOnDeviceConfigurationProgressUpdate" body:@{@"progress": [NSNumber numberWithFloat:progress]}];
+}
+
+- (void)swiper:(BMSSwiperController *)swiper displayMessage:(NSString *)message canCancel:(BOOL)cancelable {
+
+    // special message we will convert into various events
+    if ([message isEqualToString:@"PLEASE SWIPE,\nTAP, OR INSERT"]) {
+
+        // device will automatically activate when connecting, so we silently cancel the transaction
+        if (self.isConnecting) {
+            __weak RNCardConnectReactLibrary *weakSelf = self;
+
+            // This is required to give the terminal enough time to finish starting before canceling -- found in their example project
+            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
```

Comment:
This is a little funky, but I have no thoughts on a better way to handle it, and following the example is probably a good call.

Reply:
there's a lot about the connection process that didn't pass the 👃 test